### PR TITLE
Blogging prompts: Refinements

### PR DESCRIPTION
**Pull request by creativecoder**
Addresses feedback from https://github.com/Automattic/jetpack/pull/26680#pullrequestreview-1188567753

#### Changes proposed in this Pull Request:

- Adds nonce exception comments
- Use proper functions for time and date generation
- Remove locale conversion function (and tests) in favor of wpcom API change
- Remove unneeded escaping work-around for JSON data
- Better escaping/error handling in a couple of places

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

See #26680 for context

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

No user facing changes from #26680. Testing instructions from that PR
**Jetpack Site**

- Be sure the site is connected to WP.com to load Jetpack blocks in the editor
- Enable experimental blocks by adding `define( 'JETPACK_EXPERIMENTAL_BLOCKS', true )` (or Settings > Jetpack Constants if you're using jurassic.ninja)
- Set the site as "having a blog" by doing at least one of the following
    - Set the `site_intent` option to `write`
    - Set the front page to show posts in Settings > Reading
    - Set a static page for the posts page in Settings > Reading
- Create a new post, you should see a writing prompt in the placeholder for the first paragraph
- Create a new post and add the `answer_prompt` query param with the prompt id for today as the value (e.g. `/wp-admin/post-new.php?answer_prompt=2130`), and you should see a writing prompt as a pullquote block, the tag `dailyprompt` added to the post, and the `_jetpack_blogging_prompt_key` post meta set to the prompt id.
- Create a new page or custom post type, and you should _not_ see the writing prompt
- Open an existing post, and you should _not_ see the writing prompt

**Simple Site**

- Sandbox a Simple site domain
- Apply this PR to your sandbox: `bin/jetpack-downloader test jetpack try/paragraph-block-writing-prompts`
- Continue with the testing instructions, as above.

### CODE REVIEW

File: projects/plugins/jetpack/_inc/blogging-prompts.php [hunk 1]
```diff
@@ -73,7 +72,7 @@ function jetpack_get_daily_blogging_prompts( $time = 0 ) {
 	}
 
 	$blog_id = \Jetpack_Options::get_option( 'id' );
-	$path    = '/sites/' . $blog_id . '/blogging-prompts?from=' . $day_before . '&number=10&_locale=' . $locale;
+	$path    = rawurldecode( '/sites/' . $blog_id . '/blogging-prompts?from=' . $day_before . '&number=10&_locale=' . $locale );
```
Comment (anomiex) [inline projects/plugins/jetpack/_inc/blogging-prompts.php]: Wrong function, and the right function won't do the right thing as you're doing it here.
```suggestion
	$path    = '/sites/' . rawurlencode( $blog_id ) . '/blogging-prompts?from=' . rawurlencode( $day_before ) . '&number=10&_locale=' . rawurlencode( $locale );
```

File: projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php [hunk 2]
```diff
@@ -43,11 +43,14 @@ function inject_blogging_prompts() {
...
 	if ( should_load_blogging_prompts() ) {
-		$daily_prompts = wp_json_encode( jetpack_get_daily_blogging_prompts() );
+		$daily_prompts = jetpack_get_daily_blogging_prompts();
 
-		// See p7H4VZ-2cf-p2 for why prompt data is escaped this way.
 		if ( $daily_prompts ) {
-			wp_add_inline_script( 'jetpack-blocks-editor', 'var Jetpack_BloggingPrompts = JSON.parse( decodeURIComponent( "' . rawurlencode( $daily_prompts ) . '" ) );', 'before' );
+			wp_add_inline_script(
+				'jetpack-blocks-editor',
+				'var Jetpack_BloggingPrompts = ' . wp_json_encode( $daily_prompts, JSON_HEX_TAG ) . ';',
```
Comment (anomiex) [inline projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php]: If you want to be extra safe
```suggestion
				'var Jetpack_BloggingPrompts = ' . wp_json_encode( $daily_prompts, JSON_HEX_TAG | JSON_HEX_AMP ) . ';',
```

File: projects/plugins/jetpack/_inc/blogging-prompts.php [hunk 3]
```diff
@@ -32,7 +32,7 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
  * @return void
  */
 function jetpack_setup_blogging_prompt_response( $post_id ) {
-	// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+	// phpcs:ignore WordPress.Security.NonceVerification.Recommended -- Clicking a prompt response link can happen from notifications, Calypso, wp-admin, etc and only setups up a response post (tag, meta, prompt text); the user must take action to actually publish the post.
```
Comment (anomiex) [inline projects/plugins/jetpack/_inc/blogging-prompts.php]: Should we worry about the list of "draft" posts getting filled up by an attacker?

**Review by anomiex — COMMENTED** [review]
Comment on line 75 is the only one preventing me giving an approval. The others are optional as far as I'm concerned or are just for discussion.

File: projects/plugins/jetpack/_inc/blogging-prompts.php (hunk 1 above)
Comment (creativecoder) [reply projects/plugins/jetpack/_inc/blogging-prompts.php]: 🤦🏼 Working too quickly on that one--thanks for the correction.

File: projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php (hunk 2 above)
Comment (creativecoder) [reply projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php]: Nice!

File: projects/plugins/jetpack/_inc/blogging-prompts.php (hunk 3 above)
Comment (creativecoder) [reply projects/plugins/jetpack/_inc/blogging-prompts.php]: It's a fair question! In practice, I'm personally not worried--it seems like a lot of effort (or patience) to get a user to open the answer_prompt link enough times to be concerning. I think there are probably easier and lower effort attacks.

For me the desire to make answering a writing prompt as seamless as possible from multiple contexts outweighs the (arguably) minimal security benefit of a nonce in this particular case.

**Comment by creativecoder** [comment]
FYI, I'm attempting to move the mag16 locale logic to wpcom here: D93225-code

File: projects/plugins/jetpack/_inc/blogging-prompts.php (hunk 3 above)
Comment (anomiex) [reply projects/plugins/jetpack/_inc/blogging-prompts.php]: Fair enough. I don't know enough about this to second-guess the decision, I just wanted to confirm that the decision was made with relevant factors having been considered.

**Review by anomiex — APPROVED** [review]

**Pull request merged**